### DISCUSSION TITLE
Extend hide-mouse-while-typing behavior (e.g. arrow keys, backspace)

### DISCUSSION THREAD

**Discussion by ddogfoodd** [ghostty-org/ghostty#2538] [Feature Requests, Ideas]
`mouse-hide-while-typing = true` is a great feature.
When typing it hides the mouse cursor (not the terminal cursor) which makes the terminal look way cleaner. This is especially great for TUIs e.g. Neovim.

However, currently this is based on whether or not a key with a corresponding UTF-8 char is typed.
Instead I would like it extended to also hide the mouse cursor when using arrow keys, or keys like backspace / delete.

What do you think?

**Comment by mitchellh** [comment]
I converted this to a discussion since it is a suggestion rather than an accepted feature. 

Hm, I can see the desire. I think a more interesting design approach might be to hide the mouse if it contains text or if key encoding results in non-empty data sent to the pty. This latter condition would allow arrow keys as you desire to hide the mouse since they encode. 

I have a memory that I tried this before but I can't fully remember if that's true or why I didn't continue with it.

**Reply by ddogfoodd** [reply]
Sounds good, I might look into it

**Comment by edrozenberg** [comment]
A little more context for MacOS:

The pointer hides on motion key press (arrows, pg up/dn, home/end) in many "well-known" editors and many other apps for ex: Apple Pages, TextMate, BBEdit, Safari, etc.

Pointer generally hides on modified motion key press as well (Ctrl or Opt + arrows etc).

In Apple Numbers (spreadsheets), the pointer does not hide for arrow motion, maybe intentional for spreadsheet use-cases, not sure.

Function keys: in Mac GUI apps these generally do not hide the pointer, but they DO in Terminal.app. Believe this would be beneficial in a terminal app since function keys are fairly frequently used in tui apps (Midnight Commander etc) and also frequently useful in bindings in editors such as Helix for motion, editing, or other app actions.

There are here & there "funny" exceptions (some probably bugs/unintentional), for ex. TextMate does not hide the pointer for Ctrl+ Left/Right arrow. Safari does not hide the pointer on Pg Up if the page is already at the top, or for Pg Dn if the page is already at the bottom, but DOES hide the pointer for up arrow at the top and down arrow at the bottom (huh).

On Mac, the pointer generally does not hide on pressing modifiers alone (S/C/A/Cmd).

Despite some of these inconsistencies, believe it makes sense to hide the pointer for motion keys and function keys. And for any modifier combos (S/C/A/Cmd in any mix) with motion keys and function keys as well, because there are many common bindings and custom bindings that use modified motion keys and function keys for moving in modified ways (for ex. Cmd + L/R to move to the beginning/end of a line) or for various app actions. More simply, the pointer should hide if a motion key or function key is pressed either alone or as part of any combo.

It should also hide for Backspace & Delete ofc, bc these are actual typing/editing keys.

Benefit - when using shell, editors, or tui apps, it's a normal desire to get rid of the pointer blocking the user's view, when the user uses any motion keys and in some apps function keys as well. In some apps (tui file managers, tui info apps for ex), the motion keys frequently dominate usage, and the mouse pointer should go away when using these apps in this common mode.

Even if "hide pointer" for motion keys and function keys is not the Ghostty default, would abs be great to have the ability to enable this.

**Reply by ddogfoodd** [reply]
@edrozenberg 

If you want to help me, you could:

1. Open Ghostty
2. Open inspector ( if you don't know how, look it up by running `ghostty +list-keybinds` and looking for the `inspector:toggle` keybind, on Linux it's ctrl + shift + i )
3. In Inspector switch over to the Keyboard tab
4. In Ghostty and press a function key
5. In Inspector see if the Inspector shows the `Press: <some function key or something>` event
6. Click on the `Press: ...` event mentioned above
7. Look if there is a value in `Encoding to Pty` or if it says `(no data)` and report back here

You could do the same for motion keys just to be sure but I assume they probably work the same on Linux and MacOS.

**Reply by ddogfoodd** [reply]
If you know how to build from source, you could even checkout the `hide-mouse` branch of my fork (https://github.com/ddogfoodd/ghostty/tree/hide-mouse), build it and then test if it works on MacOS and fits your use cases.

**Comment by ddogfoodd** [comment]
I looked back at my approach in #2544 with shame and found a slightly better solution that might still not be optimal, but that I wanted to discuss.
As @mitchellh mentioned, while the previous approach might work, it's not how it should be designed.

For the previous approach, I found inspiration in [this part of the code](https://github.com/ghostty-org/ghostty/blob/c3d65d3975f91773f10fb19e10911e0f4d6463aa/src/Surface.zig#L1931-L1934).
However (and this is why I'm embarrassed now ...), I focused solely on the if-condition, whereas the actual important thing to note here was the position in the function. At that point we would have returned if there was nothing to encode and send to the pty.

Because we know that there was some encoding generated to send to the pty at this point, we might move the hide mouse code here.
Also, we don't need to check for utf-8 content anymore, as the pty encoding condition is more inclusive and makes other checks as utf-8 content or modifier key checks obsolete.
([See this commit in my fork](https://github.com/ddogfoodd/ghostty/commit/f7bed760b40c0eda0108d8ba0e07dac077a38421))

This seems to work (at least on Linux). I wonder how it affects performance, but am not sure how to test yet. There might also be an even better design than just moving the code a bit. And if there are no key release events that encode something to be sent to pty, we can remove one more check as well.

Any discussion is welcome.
This is still just a proposal and it's not decided if this will become an issue or a PR sometime, as not that many users seem to be interested in this so far. I still wanted to post about it for potential discussion.

**Comment by mitchellh** [comment]
This should be resolved on nightly/tip releases. We only hide the mouse on various conditions:

```zig
    // If this input event has text, then we hide the mouse if configured.
    // We only do this on pressed events to avoid hiding the mouse when we
    // change focus due to a keybinding (i.e. switching tabs).
    if (self.config.mouse_hide_while_typing and
        event.action == .press and
        !self.mouse.hidden and
        event.utf8.len > 0)
    {
        self.hideMouse();
    }
```

**Reply by ddogfoodd** [reply]
Just so I understand, is this closed as not planned?
Otherwise, this is not yet resolved on tip. The last condition `event.utf8.len > 0` still doesn't let arrow keys or backspace hide the mouse.
